File: jaros/state/machine.py
"""Transition enforcement and atomic commit (EXT-002 / REQ-2, REQ-3).

The machine permits *only* transitions declared in :mod:`jaros.state.model`.
Any other ``(state, event)`` pair is rejected with a typed
:class:`UndefinedTransitionError` and causes no mutation, so the machine can
never enter an undefined state.

:func:`commit` ties enforcement to durability: it validates the transition,
durably appends it to the log, and *only then* applies it. If the append fails,
the transition is not applied — the commit is atomic (log-then-apply).
"""

from __future__ import annotations

from dataclasses import dataclass

from jaros.state.log import LogEntry, TransitionLog
from jaros.state.model import TRANSITIONS, is_state


# #EXT-002-REQ-2 Start
class UndefinedTransitionError(ValueError):
    """Raised when a ``(state, event)`` pair is not in the transition table.

    The machine performs no mutation when this is raised, preserving the
    invariant that the current state is always a declared state.
    """


class InvalidStateError(ValueError):
    """Raised when a value is not one of the declared states."""


def assert_valid_state(state: str) -> str:
    """Assert ``state`` is a declared state; return it unchanged.

    Used as an entry/exit invariant so the machine can never operate on a state
    outside the model.

    Raises:
        InvalidStateError: if ``state`` is not a declared state.
    """
    if not is_state(state):
        raise InvalidStateError(f"not a declared state: {state!r}")
    return state


def transition(state: str, event: str) -> str:
    """Return the next state for ``(state, event)`` per the model.

    Raises:
        InvalidStateError: if ``state`` is not a declared state.
        UndefinedTransitionError: if no transition is declared for the pair.
            No mutation is performed in this case.
    """
    assert_valid_state(state)
    targets = TRANSITIONS.get(state, {})
    if event not in targets:
        raise UndefinedTransitionError(
            f"no transition for state={state!r} event={event!r}"
        )
    next_state = targets[event]
    # Invariant: the table can only ever point at declared states.
    return assert_valid_state(next_state)
# #EXT-002-REQ-2 End


# #EXT-002-REQ-3 Start
@dataclass(frozen=True)
class CommitResult:
    """The outcome of a successful :func:`commit`.

    ``state`` is the new current state after the transition was durably logged
    and applied; ``index`` is the 1-based position of the appended log entry.
    """

    state: str
    index: int


def commit(log: TransitionLog, state: str, event: str) -> CommitResult:
    """Validate, durably log, then apply a transition — atomically.

    Order of operations (atomic log-then-apply):

    1. Validate the ``(state, event)`` pair against the model. An undefined
       pair raises :class:`UndefinedTransitionError` and nothing is logged or
       applied.
    2. Durably append the resulting transition to ``log`` (flush + fsync).
    3. Only if the append succeeds, return the new state as applied.

    If step 2 raises (e.g. disk/append failure), the exception propagates and
    the transition is *not* applied — the caller's state is unchanged.

    Returns:
        CommitResult: the new state and the appended entry's 1-based index.

    Raises:
        UndefinedTransitionError: for an undefined transition (no mutation).
        Exception: whatever the log's append raises on failure (no mutation).
    """
    # 1. Validate first — undefined pairs never touch the log.
    next_state = transition(state, event)

    # 2. Durably log before applying. The index is 1-based and follows the
    #    current log length.
    index = log.length() + 1
    entry = LogEntry.make(index=index, event=event, state=next_state)
    log.append(entry)  # on failure this raises and we never reach step 3

    # 3. Append succeeded -> the transition is now committed and observable.
    return CommitResult(state=next_state, index=index)
# #EXT-002-REQ-3 End

File: src/state/machine.ts
import { isState, TRANSITIONS, type Event, type State } from "./model";
import type { LogEntry, TransitionLog } from "./log";

// #EXT-002-REQ-2 Start
/**
 * Transition enforcement ([REQ-2]). Only transitions present in the explicit
 * model (`./model`) are permitted; every other `(state, event)` pair is rejected
 * with a typed {@link UndefinedTransitionError} and causes no mutation. The
 * machine can therefore never enter a state outside the declared set.
 */

/**
 * Typed error raised when `transition()` is asked for a `(state, event)` pair
 * that is not declared in {@link TRANSITIONS}. Carrying the offending pair lets
 * callers diagnose without parsing the message.
 */
export class UndefinedTransitionError extends Error {
  readonly state: State;
  readonly event: Event;

  constructor(state: State, event: Event) {
    super(`No transition declared for state "${state}" on event "${event}".`);
    this.name = "UndefinedTransitionError";
    this.state = state;
    this.event = event;
    Object.setPrototypeOf(this, UndefinedTransitionError.prototype);
  }
}

/**
 * Typed error raised when an invariant is violated: the machine is observed
 * holding a value that is not one of the declared {@link State}s. This should be
 * impossible in normal operation and indicates corruption.
 */
export class InvalidStateError extends Error {
  readonly value: unknown;

  constructor(value: unknown) {
    super(`Value "${String(value)}" is not a declared state.`);
    this.name = "InvalidStateError";
    this.value = value;
    Object.setPrototypeOf(this, InvalidStateError.prototype);
  }
}

/**
 * Invariant check: assert `value` is one of the declared states. Used on entry
 * and exit of {@link transition} so an undeclared state can never propagate.
 * Throws {@link InvalidStateError} otherwise.
 */
export function assertValidState(value: unknown): asserts value is State {
  if (!isState(value)) {
    throw new InvalidStateError(value);
  }
}

/**
 * Pure transition function. Returns the next state iff `(state, event)` exists in
 * {@link TRANSITIONS}; otherwise throws {@link UndefinedTransitionError} and
 * performs no mutation. The invariant is asserted on the incoming state and on
 * the computed next state.
 */
export function transition(state: State, event: Event): State {
  assertValidState(state);
  const next = TRANSITIONS[state][event];
  if (next === undefined) {
    throw new UndefinedTransitionError(state, event);
  }
  assertValidState(next);
  return next;
}
// #EXT-002-REQ-2 End

// #EXT-002-REQ-3 Start
/**
 * Atomic commit ([REQ-3]). Validate the transition, then append it to the
 * durable log, then apply it. The transition becomes observable only after the
 * append succeeds: if the append throws, the new state is NOT applied and the
 * error propagates — either the transition is both logged and applied, or
 * neither.
 *
 * @returns the new state together with the persisted log entry.
 */
export function commit(
  log: TransitionLog,
  state: State,
  event: Event
): { state: State; entry: LogEntry } {
  // 1. Validate (no mutation, no I/O on rejection).
  const next = transition(state, event);
  // 2. Durably append BEFORE applying. A throw here leaves caller state intact.
  const entry = log.append({ from: state, event, to: next });
  // 3. Only now is the transition observable.
  return { state: next, entry };
}
// #EXT-002-REQ-3 End
